technical/candles.py

```diff
@@ -8,11 +8,9 @@ def heikinashi(bars):
     bars['ha_close'] = (bars['open'] + bars['high'] +
                         bars['low'] + bars['close']) / 4
 
-    bars['ha_open'] = (bars['open'].shift(1) + bars['close'].shift(1)) / 2
-    bars.loc[:1, 'ha_open'] = bars['open'].values[0]
-    for x in range(2):
-        bars.loc[1:, 'ha_open'] = (
-                                          (bars['ha_open'].shift(1) + bars['ha_close'].shift(1)) / 2)[1:]
+    bars.at[0, 'ha_open'] = (bars.at[0, 'open'] + bars.at[0, 'close']) / 2
+    for x in range(1, len(bars)):
+        bars.at[x, 'ha_open'] = (bars.at[x - 1, 'ha_open'] + bars.at[x - 1, 'ha_close']) / 2
```

Comment:
What's the performance difference between the 2? 
that's especially relevant for backtesting ...


Also, are you sure that the qtpylib implementation is correct / according to the specs?
They've had indicators calculated wrongly before - so i'm always cautious as to what can be used reliably as a indicator.

Reply:
The performance differences are actually not that much. This is an example of a basic strategy, buy and sell on Heikin Ashi reversals:
```python
# --- Do not remove these libs ---
from freqtrade.strategy.interface import IStrategy
from pandas import DataFrame
# --------------------------------

from technical.candles import heikinashi

class HA(IStrategy):

    minimal_roi = {
        "0": 1,
    }

    stoploss = -0.1

    timeframe = '4h'

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        ha_df = heikinashi(dataframe)
        dataframe['ha_reversal'] = ha_df['reversal']
#        dataframe.to_csv('user_data/output/HA-test-{}.csv'.format(metadata['pair'].replace("/","-")))

        return dataframe

    def populate_buy_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[
            (
                (dataframe['ha_reversal'] == 1)
            ),
            'buy'] = 1

        return dataframe

    def populate_sell_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe.loc[
            (
                (dataframe['ha_reversal'] == -1)
            ),
            'sell'] = 1
        return dataframe
```

If I run that with the original code with --timerange=20210101-20210201 --ticker-interval 4h :
```
============================================================
|      Pair |   Buys |   Avg Profit % |   Cum Profit % |   Tot Profit USDT |   Tot Profit % |   Avg Duration |   Wins |   Draws |   Losses |
|-----------+--------+----------------+----------------+-------------------+----------------+----------------+--------+---------+----------|
|  ETH/USDT |     21 |           2.33 |          48.88 |       48.92920276 |           0.00 |       20:34:00 |      9 |       0 |       12 |
| AAVE/USDT |     25 |           1.88 |          47.09 |       47.13580639 |           0.00 |       17:46:00 |     12 |       0 |       13 |
```
And with the proposed changes:
```
============================================================
|      Pair |   Buys |   Avg Profit % |   Cum Profit % |   Tot Profit USDT |   Tot Profit % |   Avg Duration |   Wins |   Draws |   Losses |
|-----------+--------+----------------+----------------+-------------------+----------------+----------------+--------+---------+----------|
|  ETH/USDT |     21 |           2.08 |          43.67 |       43.71834751 |           0.00 |       21:20:00 |     10 |       0 |       11 |
| AAVE/USDT |     24 |           2.03 |          48.73 |       48.77638481 |           0.00 |       18:30:00 |     11 |       0 |       13 |
```

But if you take a closer look at the original code:
```python
    bars['ha_open'] = (bars['open'].shift(1) + bars['close'].shift(1)) / 2
    bars.loc[:1, 'ha_open'] = bars['open'].values[0]
    for x in range(2):
        bars.loc[1:, 'ha_open'] = (
                                          (bars['ha_open'].shift(1) + bars['ha_close'].shift(1)) / 2)[1:]
```
then it is quite clear that this can not be correct. Since HA_open is based on the previous candle, we first need to define the first candle, and then loop through the rest of the dataframe. Now first the whole dataframe is filled, then next line calculate the first record, and then a loop runs only for 2 times (not sure why only 2 loops...).

I also have my doubts about some of the indicators in QtpyLib, but as far as I can see the Heikin Ashi implementation is matching with the specifications that I can find online. Also Tradingview is same as QtpyLib.

https://school.stockcharts.com/doku.php?id=chart_analysis:heikin_ashi
- We need our first Heikin-Ashi candlestick before we can calculate future Heikin-Ashi candlesticks.
- The first Heikin-Ashi open equals the average of the open and close ((O+C)/2)
- The Heikin-Ashi Open is the average of the prior Heikin-Ashi candlestick open plus the close of the prior Heikin-Ashi candlestick

https://www.investopedia.com/terms/h/heikinashi.asp
- Use one period to create the first Heikin-Ashi (HA) candle, using the formulas. 
- Use the open and close to create the first HA open.

Both sites mention separate calculation for the 1st candle HA-high and HA-low, using the high or low of first candle. But for this no code changes are needed as for those always current candle values are used, and the results are the same then anyway. Only HA-open uses prior candle data. And then sequence can only be: calculate the first candle before looping through the rest of the dataframe.

Hope that helps. If any questions please let me know.

Reply:
with performance, i was less worried about strategy performance (this will obviously change as the calculation changes) but about the effective calculation performance.

Trying this for 107951 candles - the prior version runs for 371ms, the new runs for 2.58s ... 

i guess it'll be ok as for 500 candles the difference is from 18ms to 52ms.
While this is still a considerable amount (~3x as much) - it's not the world if you then have correct values...

Reply:
Yes, new version will be slower as each row in the dataframe needs to be iterated in the for loop. The previous did only 2 loops. I am still new to Python, any suggestions that I can try to speed it up? I did find many online discussion about for loops being slow on Python.

Reply:
it's not about python in particular, loops are slow in general (slower than vectorized operations, anyway).

i'm fine with this as it's now the correct calculation - it's just worth to mention that it will be slower - but its expected if the calculation needs the result of the previous line.

Reply:
Have been searching for some code examples for Heikin Ashi, and measured the time to calculate the candle values and reversal points for each pair (5 minute candles from 20210101 to 20210201).


This was used in the strategy to measure times for the Technical library and QtpyLib:
````python
    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        print ("Pair: {}".format(metadata['pair'].replace("/","-")))
        start = time.time()
        ha_df = heikinashi(dataframe)
        dataframe['ha_open'] = ha_df['open']
        dataframe['ha_high'] = ha_df['high']
        dataframe['ha_low'] = ha_df['low']
        dataframe['ha_close'] = ha_df['close']
        dataframe['ha_reversal'] = ha_df['reversal']
        end = time.time()
        print("Technical time = {} [s]".format(round(end - start, 4)))

        start = time.time()
        dataframe['qtpylib_ha_open'] = qtpylib.heikinashi(dataframe)['open']
        dataframe['qtpylib_ha_high'] = qtpylib.heikinashi(dataframe)['high']
        dataframe['qtpylib_ha_low'] = qtpylib.heikinashi(dataframe)['low']
        dataframe['qtpylib_ha_close'] = qtpylib.heikinashi(dataframe)['close']
        dataframe['qtpylib_ha_green_to_red'] = qtpylib.crossed_below(dataframe['qtpylib_ha_close'], dataframe['qtpylib_ha_open'])
        dataframe['qtpylib_ha_red_to_green'] = qtpylib.crossed_above(dataframe['qtpylib_ha_close'], dataframe['qtpylib_ha_open'])
        end = time.time()
        print("QtpyLib time   = {} [s]".format(round(end - start, 4)))

        dataframe['diff_ha_open'] = dataframe['ha_open'] - dataframe['qtpylib_ha_open']
        dataframe['diff_ha_high'] = dataframe['ha_high'] - dataframe['qtpylib_ha_high']
        dataframe['diff_ha_low'] = dataframe['ha_low'] - dataframe['qtpylib_ha_low']
        dataframe['diff_ha_close'] = dataframe['ha_close'] - dataframe['qtpylib_ha_close']

        dataframe.to_csv('user_data/output/HA-test-{}.csv'.format(metadata['pair'].replace("/","-")))
````

First compared the code that is just merged. The calculations for QtpyLib are the same, but running it from the technical library is much faster already:
````
Pair: BTC-USDT
Technical time = 0.1415 [s]
QtpyLib time   = 0.4258 [s]
Pair: ETH-USDT
Technical time = 0.1216 [s]
QtpyLib time   = 0.4405 [s]
Pair: LTC-USDT
Technical time = 0.1235 [s]
QtpyLib time   = 0.4241 [s]
````

I then found this: https://stackoverflow.com/a/55110393, where the for loop is replaced with a list comprehension method. And that is significantly faster. I did try some of the other suggested codes on that page, but this list comprehension method is the clear winner.
````
Pair: BTC-USDT
Technical time = 0.0576 [s]
QtpyLib time   = 0.4412 [s]
Pair: ETH-USDT
Technical time = 0.0575 [s]
QtpyLib time   = 0.4271 [s]
Pair: LTC-USDT
Technical time = 0.0581 [s]
QtpyLib time   = 0.4438 [s]
````

The code to calculate the candle data for last test:
````python
    bars['ha_close']=bars[['open','high','low','close']].mean(axis=1)

    ha_open = [ (bars.open[0] + bars.close[0]) / 2 ]
    [ ha_open.append((ha_open[x] + bars.ha_close.values[x]) / 2) \
    for x in range(0, len(bars)-1) ]
    bars['ha_open'] = ha_open

    bars['ha_high']=bars[['ha_open','ha_close','high']].max(axis=1)
    bars['ha_low']=bars[['ha_open','ha_close','low']].min(axis=1)
````
 Would you like me to submit a new pull request?

Reply:
I just realized that the logic for QtpyLib in the strategy could be faster with loading the output of QrpyLib in a temporary dataframe. But still the list comprehension method beats it.
````python
        start = time.time()
        qt_ha_df = qtpylib.heikinashi(dataframe)
        dataframe['qtpylib_ha_open'] = qt_ha_df['open']
        dataframe['qtpylib_ha_high'] = qt_ha_df['high']
        dataframe['qtpylib_ha_low'] = qt_ha_df['low']
        dataframe['qtpylib_ha_close'] = qt_ha_df['close']
        dataframe['qtpylib_ha_green_to_red'] = qtpylib.crossed_below(qt_ha_df['close'], qt_ha_df['open'])
        dataframe['qtpylib_ha_red_to_green'] = qtpylib.crossed_above(qt_ha_df['close'], qt_ha_df['open'])
        end = time.time()
        print("QtpyLib time   = {} [s]".format(round(end - start, 4)))
````
New test times:
````
Pair: BTC-USDT
Technical time = 0.0574 [s]
QtpyLib time   = 0.1108 [s]
Pair: ETH-USDT
Technical time = 0.0602 [s]
QtpyLib time   = 0.1112 [s]
Pair: LTC-USDT
Technical time = 0.0612 [s]
QtpyLib time   = 0.1136 [s]
````

Reply:
yes please

Does this get the same result? 
i've not used a list-comprehension this way ... so please also include a comment that says something around `can't assign the list-comprehension directly, as the prior element for each loop needs to be used`

Reply:
It does give the same result. I had to look a few times before I understood what was happening with that code.

I think I will rewrite it like this to make it easier to understand:
````python
    ha_open = [ (bars.open[0] + bars.close[0]) / 2 ]
    [ ha_open.append((ha_open[x] + bars.ha_close[x]) / 2) for x in range(0, len(bars)-1) ]
    bars['ha_open'] = ha_open
````
Simplified this is: `ha_open = [calculate first record][list comprehension method loop for remaining records]`
So the first record is calculated as: `(bars.open[0] + bars.close[0]) / 2`
The following records are appended in a loop: `ha_open[x] + bars.ha_close[x]) / 2`
This means that result is this:
````
ha_open[0] =  (bars.open[0] + bars.close[0]) / 2
ha_open[1] = (ha_open[0] + bars.ha_close[0]) / 2
...
ha_open[last] = ha_open[len(bars)-1] + bars.ha_close[len(bars)-1]) / 2
````
And that is indeed how it should be.

Reply:
i understood the code after reading it trice... ;) 

i'm just not sure i (or future readers) will understand it as well :D

Reply:
I have updated my live trader with the proposed changes last night. All working fine.
Created new pull request: https://github.com/freqtrade/technical/pull/159